src/vset.c

```diff
@@ -0,0 +1,1249 @@
+ *
+ * Return:
+ *   true in case a removal was made, false otherwise */
+bool pvRemove(pVector **psv, void *elem) {
+    pVector *sv = *psv;
+    if (!sv || sv->len == 0) return false;
+
+    for (uint32_t i = 0; i < sv->len; i++) {
+        if (sv->data[i] == elem) {
+            *psv = pvRemoveAt(sv, i);
+            return true;
+        }
+    }
+    return false;
+}
+
+/* Retrieves the element at the specified index in the pVector.
+ *
+ * Arguments:
+ *   vec - The pVector to retrieve from.
+ *   idx - The index of the element to access.
+ *
+ * Return:
+ *   A pointer to the element at the given index.
+ *   Returns NULL if the vector is NULL or the index is out of bounds. */
+void *pvGet(pVector *vec, uint32_t idx) {
+    if (!vec || idx >= vec->len) return NULL;
+    return vec->data[idx];
+}
+
+/* Frees the memory used by the pVector.
+ *
+ * Arguments:
+ *   sv - The pVector to free.
+ *
+ * Return:
+ *   None. */
+void pvFree(pVector *sv) {
+    if (sv) zfree(sv);
+}
+
+uint32_t pvFind(pVector *sv, void *elem) {
+    if (!sv || sv->len == 0) return 0;
+
+    for (uint32_t i = 0; i < sv->len; i++) {
+        if (sv->data[i] == elem) {
+            return i;
+        }
+    }
+    return sv->len;
+}
+/*************************************************************************************************************
+ *                                pVector End
+ *************************************************************************************************************/
+#define VSET_BUCKET_NONE -1      // matching the NULL case
+#define VSET_BUCKET_SINGLE 0x1UL // xx1 (assuming sds)
+#define VSET_BUCKET_VECTOR 0x2UL // 010
+#define VSET_BUCKET_HT 0x4UL     // 100
+#define VSET_BUCKET_RAX 0x6UL    // 110
+
+#define VSET_TAG_MASK 0x7UL
+#define VSET_PTR_MASK (~VSET_TAG_MASK)
+
+// Determine bucket type
+static inline int vsetBucketType(vsetBucket *b) {
+    if (b == NULL) return VSET_BUCKET_NONE;
+
+    uintptr_t bits = (uintptr_t)b;
+    if (bits & 0x1)
+        return VSET_BUCKET_SINGLE;
+    return bits & VSET_TAG_MASK;
+}
+
+// Access raw pointer
+static inline void *vsetBucketRawPtr(vsetBucket *b) {
+    return (void *)((uintptr_t)b & VSET_PTR_MASK);
+}
+
+// Accessors with type assertions
+static inline pVector *vsetBucketVector(vsetBucket *b) {
+    assert(vsetBucketType(b) == VSET_BUCKET_VECTOR);
+    return (pVector *)vsetBucketRawPtr(b);
+}
+
+static inline hashtable *vsetBucketHashtable(vsetBucket *b) {
+    assert(vsetBucketType(b) == VSET_BUCKET_HT);
+    return (hashtable *)vsetBucketRawPtr(b);
+}
+
+static inline rax *vsetBucketRax(vsetBucket *b) {
+    assert(vsetBucketType(b) == VSET_BUCKET_RAX);
+    return (rax *)vsetBucketRawPtr(b);
+}
+
+static inline void *vsetBucketSingle(vsetBucket *b) {
+    return b;
+}
+
+// Setters
+static inline vsetBucket *vsetBucketSetType(vsetBucket *b, int type) {
+    uintptr_t p = (uintptr_t)b;
+    return (vsetBucket *)(p | (type & VSET_TAG_MASK));
+}
```

Comment:
Can we use the type vsetBucket only for encoded bucket pointers? The raw pointers are pVector, rax, etc. Thus, the `b` parameter shouldn't be of type vsetBucket but `void *`, which is also what vsetBucketRawPtr returns.

I'm not sure I would call it a setter. I would call it encode or maybe vsetBucketFromRawPtr or vsetBucketEncodeRawPtr.

```suggestion
static inline vsetBucket *vsetBucketFromRawPtr(void *b, int type) {
    uintptr_t p = (uintptr_t)b;
    return (vsetBucket *)(p | (type & VSET_TAG_MASK));
}
```